pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -837,6 +862,47 @@ func (c *Controller) updateEDS(ep *v1.Endpoints, event model.Event) {
 	_ = c.XDSUpdater.EDSUpdate(c.ClusterID, string(hostname), ep.Namespace, endpoints)
 }
 
+// called when the overall mTLS ready status of endpoints have changed for a service and the cached Service object must be updated
+func (c *Controller) updateSvcMtlsReady(svc *model.Service) {
+	// TODO gihanson does the mTLS ready state impact external name services?
+	c.RLock()
+	instances := c.externalNameSvcInstanceMap[svc.Hostname]
+	c.RUnlock()
+	if instances != nil {
+		return
+	}
+
+	item, exists, err := c.endpoints.informer.GetStore().GetByKey(kube.KeyFunc(svc.Attributes.Name, svc.Attributes.Namespace))
```

Comment:
why we have to iterate endpoints again here? I saw `updateEndpoints` code above, where we already process the mtls label information on each endpoint?

Reply:
@incfly my understanding is that in `updateEDS()` we are iterating over a subset of endpoints - i.e. not all the instances of a service, just the new/updated ones.

Reply:
`eds.go` `updateServiceShards` already iterate through all the endpoints, and creating `IstioEndpoint` struct. At that line, you set `MTLSReady`. We don't need to re-iterate again the endpoints.

The algorithm should be

- The initial reconcilation (listing the endpoints for a service, in `updateServiceShards`), for each endpoint, we check its mtlsReady == true or not.
- In the end of that iteration, `updateServiceShards`, we associate an integer counter to the `Service`, MTLSReadyCounter. 
- For each individual endpoints updates, check the endpoint's  mTLSReady. And update the MTLSReadyCounter.
- Service's MTLSReady = true when len(endpoints) == MTLSReadyCounter.
- If a service's MTLSReady toggle, trigger a full CDS Push.

With this algorithm, I would expect we don't have to see any new for loop.

Reply:
By the way, look at service account implementation might be helpful.

 @rshriram actually service account implementation is very similar to the problems here. That it's tied for each endpoints, and a CDS push needs to be triggered when the set(endpoints_service_accounts) change. Here the criteria is len(mtls_ready_endpoints == len(all_endpoints)).

Reply:
We do a full push when an endpoint's service accounts change. How frequently do service accounts change at runtime? Not that much.
The algorithm in this PR is triggering a full push when mtls labels change. equally costly. But you dont know which label changed. You just know that some label changed. You then iterate call eds code for any label change, check if label is present and if not present, you trigger full push. If present, you still trigger full push. Coz you dont know if the label was added/removed.

Reply:
I agree the algorithm *has to* be improved as proposed above, but I believe after that, all concerns can be addressed. This is basically the extension of our incremental EDS updates, not much different from that.

See inlined response.


> The algorithm in this PR is triggering a full push when mtls labels change. equally costly. 

See my algorithm above. We trigger push only when mTLS labels changes accumulated to all endpoints for a service, not on single label/endpoint event.

> But you dont know which label changed. You just know that some label changed. You then iterate call eds code for any label change, check if label is present and if not present

My algorithm above will not add any `for-loop` at all. A service has as counter to record the number of the endpoints with mtlsReady label. Endpoint is attached with boolean of `mTLSReady`. All updates are incremental, not iterating from scratch, which is wasteful as you points out.

> you trigger full push. If present, you still trigger full push. Coz you dont know if the label was added/removed.

I have a service with 100 endpoints. I use Istio incrementally for some subset deployments.
mTLSReady label event on 0, 1, 2... 99 pods/endpoints won't trigger any CDS push, only the 100 -> 99 and 99 -> 100 trigger the push. Other most of time, we are just maintaining the state of MTLSReady information for both service and endpoints.